File: src/main/java/gg/ethereallabs/blockChess/game/Game.java
package gg.ethereallabs.blockChess.game;

import com.github.bhlangonijr.chesslib.*;
import com.github.bhlangonijr.chesslib.move.Move;
import com.github.bhlangonijr.chesslib.move.MoveList;
import gg.ethereallabs.blockChess.BlockChess;
import gg.ethereallabs.blockChess.config.Config;
import gg.ethereallabs.blockChess.data.LocalStorage;
import gg.ethereallabs.blockChess.data.PlayerData;
import gg.ethereallabs.blockChess.elo.EloManager;
import gg.ethereallabs.blockChess.engine.EnemyData;
import gg.ethereallabs.blockChess.engine.UciEngine;
import gg.ethereallabs.blockChess.gui.GameGUI;
import gg.ethereallabs.blockChess.utils.SyncHelper;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.event.ClickEvent;
import net.kyori.adventure.text.event.HoverEvent;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.Sound;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class Game {

    public final Board board = new Board();
    public Player white = null;
    public Player black = null;
    public Player drawRequester = null;
    public MoveList moveList = new MoveList();
    public List<Piece> whiteEaten = new ArrayList<>();
    public List<Piece> blackEaten = new ArrayList<>();
    private boolean ended = false;

    public Move lastMove = null;

    private int taskId = -1;
    public long whiteTimeMs = 5L * 60 * 1000;
    public long blackTimeMs = 5L * 60 * 1000;
    private long lastTickMs = System.currentTimeMillis();

    public GameGUI guiWhite = null;
    public GameGUI guiBlack = null;

    public boolean againstBot = false;
    private Side engineSide = null;
    private UciEngine engine = null;
    private boolean engineThinking = false;
    private Integer engineSkill = null;
    public EnemyData botData = null;

    private ResultType matchResult = null;

    public enum ResultType {
        WHITE_WIN, BLACK_WIN,
        DRAW_STALEMATE, DRAW_REPETITION, DRAW_INSUFFICIENT, DRAW_100MOVES,
        TIMEOUT_WHITE, TIMEOUT_BLACK,
        WHITE_RESIGN, BLACK_RESIGN, DRAW,
        WHITE_FORFEIT, BLACK_FORFEIT
    }

    public boolean isEnded() { return ended; }

    // ─── GAME START ───────────────────────────────────────────────────────────

    public void start(Player pWhite, Player pBlack) {
        setupBoard();
        white = pWhite;
        black = pBlack;
        guiWhite = new GameGUI(this, true);
        guiBlack = new GameGUI(this, false);
        guiWhite.open(pWhite);
        guiBlack.open(pBlack);
        startTimer();
    }

    public void startAgainstBot(Player human, int difficulty, boolean humanIsWhite) {
        if (!Config.botEnabled) {
            BlockChess.instance.sendMessage("<red>Bots are currently disabled!", human);
            human.playSound(human.getLocation(), Sound.ENTITY_VILLAGER_NO, 1f, 1f);
            return;
        }

        setupBoard();
        againstBot = true;
        engineSkill = difficulty;
        engineSide = humanIsWhite ? Side.BLACK : Side.WHITE;

        if (humanIsWhite) {
            white = human;
            guiWhite = new GameGUI(this, true);
            guiWhite.open(human);
        } else {
            black = human;
            guiBlack = new GameGUI(this, false);
            guiBlack.open(human);
        }

        engine = new UciEngine(Config.enginePath);
        try {
            engine.start();
            engine.initLevel(engineSkill);
        } catch (Exception ex) {
            human.sendMessage(BlockChess.mm.deserialize("<red>Impossible to start chess engine: " + ex.getMessage()));
        }

        startTimer();

        if (board.getSideToMove() == engineSide) triggerEngineMove();
    }

    private void setupBoard() {
        board.loadFromFen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1");
    }

    // ─── GAME END ─────────────────────────────────────────────────────────────

    public void finalizeGame(ResultType result) {
        if (ended) return;
        ended = true;
        stop();

        matchResult = result;
        sendPGN();

        if (againstBot) {
            handleBotResult(result);
        } else {
            handlePlayerResult(result);
        }

        GameManager.end(this);
    }

    private void sendPGN() {
        String pgn = moveList.toString();
        Component pgnMsg = Component.text("PGN (Click to copy): ", NamedTextColor.GRAY)
                .append(
                    Component.text(pgn, NamedTextColor.YELLOW)
                        .clickEvent(ClickEvent.copyToClipboard(pgn))
                        .hoverEvent(HoverEvent.showText(Component.text("Click to copy PGN")))
                );
        if (white != null && white.isOnline()) white.sendMessage(pgnMsg);
        if (black != null && black.isOnline()) black.sendMessage(pgnMsg);
    }

    private void handlePlayerResult(ResultType result) {
        if (white == null || black == null) return;
        Player w = white;
        Player b = black;

        PlayerData wData = EloManager.players.computeIfAbsent(w.getUniqueId(), k -> new PlayerData());
        PlayerData bData = EloManager.players.computeIfAbsent(b.getUniqueId(), k -> new PlayerData());

        switch (result) {
            case WHITE_WIN    -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by Checkmate");
            case BLACK_WIN    -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by Checkmate");
            case TIMEOUT_BLACK -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by time");
            case TIMEOUT_WHITE -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by time");
            case WHITE_RESIGN  -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by resignation");
            case BLACK_RESIGN  -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by resignation");
            case WHITE_FORFEIT -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by abandonment");
            case BLACK_FORFEIT -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by abandonment");
            case DRAW_STALEMATE  -> announceDraw(w, b, wData, bData, "⚖", "by stalemate");
            case DRAW_REPETITION -> announceDraw(w, b, wData, bData, "🔁", "by repetition");
            case DRAW_INSUFFICIENT -> announceDraw(w, b, wData, bData, "🪶", "by insufficient material");
            case DRAW_100MOVES   -> announceDraw(w, b, wData, bData, "⏳", "by 50 move rule");
            case DRAW            -> announceDraw(w, b, wData, bData, "⚖", "by agreed draw");
        }

        LocalStorage.savePlayerData(w);
        LocalStorage.savePlayerData(b);
    }

    private void handleBotResult(ResultType result) {
        Player human = white != null ? white : black;
        if (human == null) return;
        boolean humanIsWhite = human == white;

        switch (result) {
            case WHITE_WIN  -> sendHumanGameMessage(human, humanIsWhite,  "🏆", "❌", "by Checkmate");
            case BLACK_WIN  -> sendHumanGameMessage(human, !humanIsWhite, "🏆", "❌", "by Checkmate");
            case TIMEOUT_WHITE -> sendHumanGameMessage(human, !humanIsWhite, "⏳", "⏳", "by time");
            case TIMEOUT_BLACK -> sendHumanGameMessage(human, humanIsWhite,  "⏳", "⏳", "by time");
            case WHITE_RESIGN, BLACK_RESIGN -> sendHumanGameMessage(human, false, "❌", "❌", "by resignation");
            case WHITE_FORFEIT, BLACK_FORFEIT -> sendHumanGameMessage(human, false, "❌", "❌", "by abandonment");
            case DRAW_STALEMATE  -> sendDrawMessage(human, "⚖", "by stalemate");
            case DRAW_REPETITION -> sendDrawMessage(human, "🔁", "by repetition");
            case DRAW_INSUFFICIENT -> sendDrawMessage(human, "🪶", "by insufficient material");
            case DRAW_100MOVES   -> sendDrawMessage(human, "⏳", "by 50 move rule");
            case DRAW            -> sendDrawMessage(human, "⚖", "by agreed draw");
        }
    }

    private void sendHumanGameMessage(Player human, boolean won, String wEmoji, String lEmoji, String reason) {
        if (!human.isOnline()) return;
        String botName = botData != null ? botData.color() + botData.name() : "Bot";
        if (won) {
            BlockChess.instance.sendMessage("<bold><#f3fa6b>" + wEmoji + "</bold> You won against " + botName + " <gray>(" + reason + ")", human);
        } else {
            BlockChess.instance.sendMessage("<bold><red>" + lEmoji + "</bold> You lost against " + botName + " <gray>(" + reason + ")", human);
        }
    }

    private void sendDrawMessage(Player human, String emoji, String reason) {
        if (!human.isOnline()) return;
        human.sendMessage(BlockChess.mm.deserialize("<yellow>" + emoji + " " + reason + "</yellow>"));
    }

    // ─── ELO / MESSAGE HELPERS ────────────────────────────────────────────────

    private void announceResult(Player winner, Player loser, PlayerData wData, PlayerData lData,
                                 double wScore, double lScore, String wEmoji, String lEmoji, String cause) {
        int wGain = EloManager.updateElo(wData, lData, wScore);
        int bGain = EloManager.updateElo(lData, wData, lScore);
        wData.wins++;
        lData.losses++;
        String wGainStr = wGain >= 0 ? "+" + wGain : String.valueOf(wGain);
        String bGainStr = bGain >= 0 ? "+" + bGain : String.valueOf(bGain);
        String wName = EloManager.getChessistName(winner);
        String lName = EloManager.getChessistName(loser);
        int wElo = wData.rating;
        int lElo = lData.rating;

        if (winner.isOnline()) {
            BlockChess.instance.sendMessage("<bold><#f3fa6b>" + wEmoji + "</bold> " + wName + " <#faff9c>(" + wElo + ") <#97ff82>" + wGainStr + " <gray>- " + lName + " <#faff9c>(" + lElo + ")<red> " + bGainStr, winner);
            BlockChess.instance.sendMessage("<bold>You Won", winner);
            BlockChess.instance.sendMessage("<gray>" + cause, winner);
        }

        if (loser.isOnline()) {
            BlockChess.instance.sendMessage("<bold><red>" + lEmoji + "</bold> " + lName + " <#faff9c>(" + lElo + ") <red>" + bGainStr + " <gray>- " + wName + " <#faff9c>(" + wElo + ")<#97ff82> " + wGainStr, loser);
            BlockChess.instance.sendMessage("<bold>You Lost", loser);
            BlockChess.instance.sendMessage("<gray>" + cause, loser);
        }
    }

    private void announceDraw(Player w, Player b, PlayerData wData, PlayerData bData, String emoji, String cause) {
        int wGain = EloManager.updateElo(wData, bData, 0.5);
        int bGain = EloManager.updateElo(bData, wData, 0.5);
        wData.draws++;
        bData.draws++;

        String wName = EloManager.getChessistName(w);
        String bName = EloManager.getChessistName(b);
        int wElo = wData.rating;
        int bElo = bData.rating;
        String wGainStr = wGain >= 0 ? "+" + wGain : String.valueOf(wGain);
        String bGainStr = bGain >= 0 ? "+" + bGain : String.valueOf(bGain);

        if (w.isOnline()) {
            BlockChess.instance.sendMessage("<#f3fa6b>" + emoji + " " + wName + " <#faff9c>(" + wElo + ") <gray>" + wGainStr + " - " + bName + " <#faff9c>(" + bElo + ") <gray>" + bGainStr, w);
            BlockChess.instance.sendMessage("<gray>" + cause, w);
        }

        if (b.isOnline()) {
            BlockChess.instance.sendMessage("<#f3fa6b>" + emoji + " " + bName + " <#faff9c>(" + bElo + ") <gray>" + bGainStr + " - " + wName + " <#faff9c>(" + wElo + ") <gray>" + wGainStr, b);
            BlockChess.instance.sendMessage("<gray>" + cause, b);
        }
    }

    // ─── TIMER & MOVE HANDLING ────────────────────────────────────────────────

    public void stop() {
        if (taskId != -1) Bukkit.getScheduler().cancelTask(taskId);
        taskId = -1;
        if (againstBot && engine != null) {
            try { engine.stop(); } catch (Exception ignored) {}
        }
    }

    private void startTimer() {
        lastTickMs = System.currentTimeMillis();
        taskId = Bukkit.getScheduler().scheduleSyncRepeatingTask(BlockChess.instance, () -> {
            long now = System.currentTimeMillis();
            long delta = now - lastTickMs;
            lastTickMs = now;

            if (board.getSideToMove() == Side.WHITE) whiteTimeMs -= delta;
            else blackTimeMs -= delta;

            if (whiteTimeMs <= 0 || blackTimeMs <= 0) {
                finalizeGame(whiteTimeMs <= 0 ? ResultType.TIMEOUT_WHITE : ResultType.TIMEOUT_BLACK);
            } else {
                if (guiWhite != null) guiWhite.updateClock();
                if (guiBlack != null) guiBlack.updateClock();
                if (againstBot && board.getSideToMove() == engineSide && !engineThinking) triggerEngineMove();
            }
        }, 20L, 20L);
    }

    public void onMoveMade(Move move) {
        lastMove = move;
        if (guiWhite != null) guiWhite.draw(white);
        if (guiBlack != null) guiBlack.draw(black);
        moveList.add(move);

        Sound sound = Sound.BLOCK_NETHER_WOOD_BREAK;
        if (board.isKingAttacked()) {
            sound = Sound.UI_BUTTON_CLICK;
        }

        Sound finalSound = sound;
        List.of(white, black).stream()
                .filter(p -> p != null && p.isOnline())
                .forEach(p -> p.playSound(p.getLocation(), finalSound, 1f, 1f));

        if (board.isStaleMate()) {
            finalizeGame(ResultType.DRAW_STALEMATE);
        } else if (board.isMated()) {
            finalizeGame(board.getSideToMove() == Side.WHITE ? ResultType.BLACK_WIN : ResultType.WHITE_WIN);
        } else if (board.isRepetition()) {
            finalizeGame(ResultType.DRAW_REPETITION);
        } else if (board.isInsufficientMaterial()) {
            finalizeGame(ResultType.DRAW_INSUFFICIENT);
        } else if (board.getHalfMoveCounter() >= 100) {
            finalizeGame(ResultType.DRAW_100MOVES);
        }
    }

    // ─── ENGINE LOGIC ─────────────────────────────────────────────────────────

    private void triggerEngineMove() {
        if (!againstBot || engine == null || engineThinking) return;
        engineThinking = true;

        String fen = board.getFen();
        long wtime = whiteTimeMs;
        long btime = blackTimeMs;

        Player human = engineSide == Side.WHITE ? black : white;
        if (human != null && human.isOnline() && botData != null) {
            BlockChess.instance.sendMessage(botData.color() + botData.name() + " <gray>is thinking...", human);
        }

        engine.positionFen(fen);

        int skill = engineSkill != null ? engineSkill : 5;

        CompletableFuture.runAsync(() -> {
            try {
                long movetime = Math.max(50, Math.min(200, skill * 50L));
                engine.goBestMoveMovetime(movetime, bestMove -> {
                    SyncHelper.runSync(() -> {
                        try {
                            Move m = uciToLegalMove(bestMove);
                            if (m != null) {
                                board.doMove(m);
                                onMoveMade(m);
                            }
                        } catch (Exception ignored) {}
                        engineThinking = false;
                    });
                });
            } catch (Exception e) {
                long alloc = Math.max(100, Math.min(2000, (board.getSideToMove() == Side.WHITE ? wtime : btime) / 20));
                engine.goBestMoveMovetime(alloc, bestMove -> {
                    SyncHelper.runSync(() -> {
                        try {
                            Move m = uciToLegalMove(bestMove);
                            if (m != null) {
                                board.doMove(m);
                                onMoveMade(m);
                            }
                        } catch (Exception ignored) {}
                        engineThinking = false;
                    });
                });
            }
        });
    }

    private Move uciToLegalMove(String uci) {
        if (uci == null || uci.isBlank() || uci.length() < 4) return null;
        Square from = Square.squareAt((uci.charAt(1) - '1') * 8 + (uci.charAt(0) - 'a'));
        Square to   = Square.squareAt((uci.charAt(3) - '1') * 8 + (uci.charAt(2) - 'a'));
        Piece promo = uci.length() >= 5 ? promoPieceFromChar(uci.charAt(4), board.getSideToMove()) : Piece.NONE;

        for (Move m : board.legalMoves()) {
            if (m.getFrom() == from && m.getTo() == to &&
                    (promo == Piece.NONE || m.getPromotion() == promo)) {
                return m;
            }
        }
        return null;
    }

    private Piece promoPieceFromChar(char c, Side side) {
        return switch (Character.toLowerCase(c)) {
            case 'q' -> side == Side.WHITE ? Piece.WHITE_QUEEN  : Piece.BLACK_QUEEN;
            case 'r' -> side == Side.WHITE ? Piece.WHITE_ROOK   : Piece.BLACK_ROOK;
            case 'b' -> side == Side.WHITE ? Piece.WHITE_BISHOP : Piece.BLACK_BISHOP;
            case 'n' -> side == Side.WHITE ? Piece.WHITE_KNIGHT : Piece.BLACK_KNIGHT;
            default  -> Piece.NONE;
        };
    }

    public GameGUI getPlayerGUI(Player player) {
        if (player == white) return guiWhite;
        if (player == black) return guiBlack;
        return null;
    }
}

File: src/main/kotlin/gg/ethereallabs/blockChess/game/Game.kt
package gg.ethereallabs.blockChess.game

import com.github.bhlangonijr.chesslib.*
import com.github.bhlangonijr.chesslib.move.Move
import com.github.bhlangonijr.chesslib.move.MoveList
import gg.ethereallabs.blockChess.BlockChess
import gg.ethereallabs.blockChess.config.Config
import gg.ethereallabs.blockChess.data.LocalStorage
import gg.ethereallabs.blockChess.elo.EloManager
import gg.ethereallabs.blockChess.data.PlayerData
import gg.ethereallabs.blockChess.engine.EnemyData
import gg.ethereallabs.blockChess.engine.UciEngine
import gg.ethereallabs.blockChess.events.ChessListener
import gg.ethereallabs.blockChess.gui.GameGUI
import gg.ethereallabs.blockChess.utils.SyncHelper.runSync
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.event.ClickEvent
import net.kyori.adventure.text.event.HoverEvent
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextColor
import org.bukkit.Bukkit
import org.bukkit.Sound
import org.bukkit.entity.Player
import java.util.concurrent.CompletableFuture.runAsync
import kotlin.collections.mutableListOf

class Game {

    val board = Board()
    var white: Player? = null
    var black: Player? = null
    var drawRequester : Player? = null
    var moveList: MoveList = MoveList()
    var whiteEaten = mutableListOf<Piece>()
    var blackEaten = mutableListOf<Piece>()
    var ended = false
        private set

    var lastMove : Move? = null

    private var taskId: Int = -1
    var whiteTimeMs: Long = 5 * 60 * 1000
    var blackTimeMs: Long = 5 * 60 * 1000
    private var lastTickMs: Long = System.currentTimeMillis()

    var guiWhite: GameGUI? = null
    var guiBlack: GameGUI? = null

    // Engine
    var againstBot: Boolean = false
    private var engineSide: Side? = null
    private var engine: UciEngine? = null
    private var engineThinking: Boolean = false
    private var engineSkill: Int? = null
    var botData : EnemyData? = null

    private var matchResult : ResultType? = null

    enum class ResultType {
        WHITE_WIN, BLACK_WIN,
        DRAW_STALEMATE, DRAW_REPETITION, DRAW_INSUFFICIENT, DRAW_100MOVES,
        TIMEOUT_WHITE, TIMEOUT_BLACK,
        WHITE_RESIGN, BLACK_RESIGN, DRAW
    }

    // ─────────────────────────────────────────────
    //               GAME START METHODS
    // ─────────────────────────────────────────────
    fun start(pWhite: Player, pBlack: Player) {
        setupBoard()
        white = pWhite
        black = pBlack
        guiWhite = GameGUI(this, true)
        guiBlack = GameGUI(this, false)
        guiWhite?.open(pWhite)
        guiBlack?.open(pBlack)
        board.addEventListener(BoardEventType.ON_MOVE, ChessListener())
        startTimer()
    }

    fun startAgainstBot(human: Player, difficulty: Int, humanIsWhite: Boolean = true) {
        setupBoard()
        againstBot = true
        engineSkill = difficulty
        engineSide = if (humanIsWhite) Side.BLACK else Side.WHITE

        if (humanIsWhite) {
            white = human
            guiWhite = GameGUI(this, true)
            guiWhite?.open(human)
        } else {
            black = human
            guiBlack = GameGUI(this, false)
            guiBlack?.open(human)
        }

        engine = UciEngine(Config.enginePath).apply {
            try {
                start()
                initLevel(engineSkill)
            } catch (ex: Exception) {
                human.sendMessage(BlockChess.mm.deserialize("<red>Impossible to start chess engine: ${ex.message}"))
            }
        }

        startTimer()

        if (board.sideToMove == engineSide) triggerEngineMove()
    }

    private fun setupBoard() {
        board.loadFromFen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1")
    }

    // ─────────────────────────────────────────────
    //                  GAME END
    // ─────────────────────────────────────────────
    fun finalizeGame(result: ResultType) {
        if (ended) return
        ended = true
        stop()

        matchResult = result

        sendPGN()

        if (againstBot) {
            handleBotResult(result)
        } else {
            handlePlayerResult(result)
        }

        GameManager.end(this)
    }

    private fun sendPGN() {
        val pgn = moveList.toString()
        val pgnMsg = Component.text("PGN (Click to copy): ", NamedTextColor.GRAY)
            .append(
                Component.text(pgn, NamedTextColor.YELLOW)
                    .clickEvent(ClickEvent.copyToClipboard(pgn))
                    .hoverEvent(HoverEvent.showText(Component.text("Click to copy PGN")))
            )
        white?.sendMessage(pgnMsg)
        black?.sendMessage(pgnMsg)
    }

    private fun handlePlayerResult(result: ResultType) {
        val w = white ?: return
        val b = black ?: return

        val wData = EloManager.players.getOrPut(w.uniqueId) { PlayerData() }
        val bData = EloManager.players.getOrPut(b.uniqueId) { PlayerData() }

        when (result) {
            ResultType.WHITE_WIN -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by Checkmate")
            ResultType.BLACK_WIN -> announceResult(b, w, wData, bData, 0.0, 1.0, "🏆", "❌", "by Checkmate")
            ResultType.TIMEOUT_BLACK -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by time")
            ResultType.TIMEOUT_WHITE -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by time")
            ResultType.WHITE_RESIGN -> announceResult(b, w, bData, wData, 1.0, 0.0, "🏆", "❌", "by resignation")
            ResultType.BLACK_RESIGN -> announceResult(w, b, wData, bData, 1.0, 0.0, "🏆", "❌", "by resignation")
            ResultType.DRAW_STALEMATE -> announceDraw(w, b, wData, bData, "⚖", "by stalemate")
            ResultType.DRAW_REPETITION -> announceDraw(w, b, wData, bData, "🔁", "by repetition")
            ResultType.DRAW_INSUFFICIENT -> announceDraw(w, b, wData, bData, "🪶", "by insufficient material")
            ResultType.DRAW_100MOVES -> announceDraw(w, b, wData, bData, "⏳", "by 50 move rule")
            ResultType.DRAW -> announceDraw(w, b, wData, bData, "⚖", "by agreed draw")
        }

        LocalStorage.savePlayerData(w)
        LocalStorage.savePlayerData(b)
    }

    private fun handleBotResult(result: ResultType) {
        val human = white ?: black ?: return
        val humanIsWhite = white == human
        val winnerIsWhite = when (result) {
            ResultType.WHITE_WIN, ResultType.TIMEOUT_BLACK, ResultType.BLACK_RESIGN -> true
            ResultType.BLACK_WIN, ResultType.TIMEOUT_WHITE, ResultType.WHITE_RESIGN -> false
            else -> null
        }

        val (winner, loser) = when (winnerIsWhite) {
            true -> Pair(humanIsWhite, !humanIsWhite)
            false -> Pair(!humanIsWhite, humanIsWhite)
            null -> Pair(null, null)
        }

        when (result) {
            ResultType.WHITE_WIN ->
                sendHumanGameMessage(human, humanIsWhite, "🏆", "❌", "by Checkmate")
            ResultType.BLACK_WIN ->
                sendHumanGameMessage(human, humanIsWhite, "🏆", "❌","by Checkmate")
            ResultType.TIMEOUT_WHITE-> sendHumanGameMessage(human, winnerIsWhite == humanIsWhite, "⏳", "⏳", "by time")
            ResultType.TIMEOUT_BLACK -> sendHumanGameMessage(human, winnerIsWhite == humanIsWhite, "⏳", "⏳", "by time")
            ResultType.WHITE_RESIGN, ResultType.BLACK_RESIGN -> sendHumanGameMessage(human, false, "❌","❌", "by resignation")
            ResultType.DRAW_STALEMATE -> sendDrawMessage(human, "⚖", "by stalemate")
            ResultType.DRAW_REPETITION -> sendDrawMessage(human, "🔁", "by repetition")
            ResultType.DRAW_INSUFFICIENT -> sendDrawMessage(human, "🪶", "by insufficient material")
            ResultType.DRAW_100MOVES -> sendDrawMessage(human, "⏳", "by 50 move rule")
            ResultType.DRAW -> sendDrawMessage(human, "⚖", "by agreed draw")
        }
    }

    private fun sendHumanGameMessage(human: Player, won: Boolean, wEmoji: String,lEmoji : String, reason: String) {
        if (won)
            BlockChess.instance.sendMessage("<bold><#f3fa6b>$wEmoji</bold> You won against ${botData?.color}${botData?.name} <gray>($reason)", human)
        else
            BlockChess.instance.sendMessage("<bold><red>$lEmoji</bold> You lost against ${botData?.color}${botData?.name} <gray>($reason)", human)
    }

    private fun sendDrawMessage(human: Player, emoji: String, reason: String) {
        human.sendMessage(BlockChess.mm.deserialize("<yellow>$emoji $reason</yellow>"))
    }

    // ─────────────────────────────────────────────
    //             ELO / MESSAGE HELPERS
    // ─────────────────────────────────────────────
    private fun announceResult(
        winner: Player,
        loser: Player,
        wData: PlayerData,
        lData: PlayerData,
        wScore: Double,
        lScore: Double,
        wEmoji: String,
        lEmoji: String,
        cause: String
    ) {
        val wGain = EloManager.updateElo(wData, lData, wScore)
        val bGain = EloManager.updateElo(lData, wData, lScore)
        wData.wins++
        lData.losses++
        val wGainStr = if (wGain >= 0) "+$wGain" else "$wGain"
        val bGainStr = if (bGain >= 0) "+$bGain" else "$bGain"
        val wName = EloManager.getChessistName(winner)
        val lName = EloManager.getChessistName(loser)
        val wElo = wData.rating
        val lElo = lData.rating
        BlockChess.instance.sendMessage("<bold><#f3fa6b>$wEmoji</bold> $wName <#faff9c>($wElo) <#97ff82>$wGainStr <gray>- $lName <#faff9c>($lElo)<red> $bGainStr", winner)
        BlockChess.instance.sendMessage("<bold><red>$lEmoji</bold> $lName <#faff9c>($lElo) <red>$bGainStr <gray>- $wName <#faff9c>($wElo)<#97ff82> $wGainStr", loser)
        BlockChess.instance.sendMessage("<bold>You Won", winner)
        BlockChess.instance.sendMessage("<bold>You Lost", loser)
        BlockChess.instance.sendMessage("<gray>$cause", winner, loser)
    }

    private fun announceDraw(w: Player, b: Player, wData: PlayerData, bData: PlayerData, emoji: String, cause: String) {
        val wGain = EloManager.updateElo(wData, bData, 0.5)
        val bGain = EloManager.updateElo(bData, wData, 0.5)
        wData.draws++
        bData.draws++

        val wName = EloManager.getChessistName(w)
        val bName = EloManager.getChessistName(b)
        val wElo = wData.rating
        val bElo = bData.rating
        val wGainStr = if (wGain >= 0) "+$wGain" else "$wGain"
        val bGainStr = if (bGain >= 0) "+$bGain" else "$bGain"

        BlockChess.instance.sendMessage("<#f3fa6b>$emoji $wName <#faff9c>($wElo) <gray>$wGainStr - $bName <#faff9c>($bElo) <gray>$bGainStr", w)
        BlockChess.instance.sendMessage("<#f3fa6b>$emoji $bName <#faff9c>($bElo) <gray>$bGainStr - $wName <#faff9c>($wElo) <gray>$wGainStr", b)
        BlockChess.instance.sendMessage("<gray>$cause", w, b)
    }

    // ─────────────────────────────────────────────
    //              TIMER & MOVE HANDLING
    // ─────────────────────────────────────────────
    fun stop() {
        if (taskId != -1) Bukkit.getScheduler().cancelTask(taskId)
        taskId = -1
        if (againstBot) try { engine?.stop() } catch (_: Exception) {}
    }

    private fun startTimer() {
        lastTickMs = System.currentTimeMillis()
        taskId = Bukkit.getScheduler().scheduleSyncRepeatingTask(BlockChess.instance, Runnable {
            val now = System.currentTimeMillis()
            val delta = now - lastTickMs
            lastTickMs = now

            if (board.sideToMove == Side.WHITE) whiteTimeMs -= delta else blackTimeMs -= delta

            if (whiteTimeMs <= 0 || blackTimeMs <= 0) {
                finalizeGame(if (whiteTimeMs <= 0) ResultType.TIMEOUT_WHITE else ResultType.TIMEOUT_BLACK)
            } else {
                guiWhite?.updateClock()
                guiBlack?.updateClock()
                if (againstBot && board.sideToMove == engineSide && !engineThinking) triggerEngineMove()
            }
        }, 20L, 20L)
    }

    fun onMoveMade(move: Move) {
        lastMove = move
        guiWhite?.draw(white)
        guiBlack?.draw(black)
        moveList.add(move)

        var sound = Sound.BLOCK_NETHER_WOOD_BREAK

        if(board.isKingAttacked){
            sound = Sound.UI_BUTTON_CLICK
        }

        listOfNotNull(white, black).forEach {
            it.playSound(it.location, sound, 1f, 1f)
        }

        when {
            board.isStaleMate -> finalizeGame(ResultType.DRAW_STALEMATE)
            board.isMated -> finalizeGame(if (board.sideToMove == Side.WHITE) ResultType.BLACK_WIN else ResultType.WHITE_WIN)
            board.isRepetition -> finalizeGame(ResultType.DRAW_REPETITION)
            board.isInsufficientMaterial -> finalizeGame(ResultType.DRAW_INSUFFICIENT)
            board.halfMoveCounter >= 100 -> finalizeGame(ResultType.DRAW_100MOVES)
        }
    }

    // ─────────────────────────────────────────────
    //                 ENGINE LOGIC
    // ─────────────────────────────────────────────
    private fun triggerEngineMove() {
        if (!againstBot || engine == null || engineThinking) return
        engineThinking = true
        val fen = board.fen
        val wtime = whiteTimeMs
        val btime = blackTimeMs
        val human = if (engineSide == Side.WHITE) black else white
        human?.let { BlockChess.instance.sendMessage("${botData?.color}${botData?.name} <gray>is thinking...", it) }

        engine!!.positionFen(fen)

        val engineCallback: (String?) -> Unit = { bestMove ->
            runSync{
                try {
                    uciToLegalMove(bestMove)?.let {
                        board.doMove(it)
                        onMoveMade(it)
                    }
                } catch (_: Exception) {}
                engineThinking = false
            }
        }

        val skill = engineSkill ?: 5

        runAsync {
            try {
                val movetime = (skill * 50L).coerceIn(50, 200)
                engine!!.goBestMoveMovetime(movetime, engineCallback)
            } catch (_: Exception) {
                val alloc = ((if (board.sideToMove == Side.WHITE) wtime else btime) / 20).coerceIn(100, 2000)
                engine!!.goBestMoveMovetime(alloc, engineCallback)
            }
        }
    }


    private fun uciToLegalMove(uci: String?): Move? {
        if (uci.isNullOrBlank() || uci.length < 4) return null
        val from = Square.squareAt((uci[1] - '1') * 8 + (uci[0] - 'a'))
        val to = Square.squareAt((uci[3] - '1') * 8 + (uci[2] - 'a'))
        val promo = if (uci.length >= 5) promoPieceFromChar(uci[4], board.sideToMove) else Piece.NONE

        return board.legalMoves().firstOrNull { it.from == from && it.to == to && (promo == Piece.NONE || it.promotion == promo) }
    }

    private fun promoPieceFromChar(c: Char, side: Side): Piece = when (c.lowercaseChar()) {
        'q' -> if (side == Side.WHITE) Piece.WHITE_QUEEN else Piece.BLACK_QUEEN
        'r' -> if (side == Side.WHITE) Piece.WHITE_ROOK else Piece.BLACK_ROOK
        'b' -> if (side == Side.WHITE) Piece.WHITE_BISHOP else Piece.BLACK_BISHOP
        'n' -> if (side == Side.WHITE) Piece.WHITE_KNIGHT else Piece.BLACK_KNIGHT
        else -> Piece.NONE
    }

    fun getPlayerGUI(player: Player) : GameGUI?{
        if(player == white)
            return guiWhite
        else if(player == black)
            return guiBlack
        else
            return null
    }
}
